File: argon-plugin/shared/src/main/scala/dev/argon/plugin/TubeLoader.scala
package dev.argon.plugin

import dev.argon.util.*
import dev.argon.io.*
import zio.*
import java.io.IOException

trait TubeLoader[Options[_, _]] {
  val supportedExtensions: Seq[String]
  
  def load[R, E](options: Options[R, E])(resource: BinaryResource[R, E]): ZIO[R & Scope, E, SerializedTube[R, E]]
}

File: plugins/api/java/src/dev/argon/plugin/api/TubeLoader.java
package dev.argon.plugin.api;

import java.io.IOException;
import java.util.List;
import dev.argon.plugin.api.resource.BinaryResource;

public interface TubeLoader<E extends Exception, TOptions> {
    List<String> supportedExtensions();    
    CloseableTube<E> load(TOptions options, BinaryResource<E> resource) throws E, IOException, PluginException;
}
